[CodeGen] Fix EVT::changeVectorElementType assertion on simple-to-extended fallback

Islam-Imad opened the pull request:
Fixes #171608

Islam-Imad committed:
[CodeGen] Fix EVT::changeVectorElementType assertion on simple-to-extended fallback
Updated ~100+ call sites across SelectionDAG and all target backends to
pass *DAG.getContext() or the available LLVMContext parameter.
Fixes #171608

topperc commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -11890,7 +11890,9 @@ SDValue TargetLowering::expandFP_ROUND(SDNode *Node, SelectionDAG &DAG) const {
     // Guillaume Melquiond. "When double rounding is odd." 17th IMACS
     // World Congress. 2005.
-    EVT F32 = VT.isVector() ? VT.changeVectorElementType(MVT::f32) : MVT::f32;
+    EVT F32 = VT.isVector()
```
Can we use `changeElementType` here?

topperc commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -11917,7 +11919,9 @@ SDValue TargetLowering::expandFP_ROUND(SDNode *Node, SelectionDAG &DAG) const {
                      DAG.getShiftAmountConstant(16, I32, dl));
     Op = DAG.getNode(ISD::BITCAST, dl, I32, Op);
-    EVT I16 = I32.isVector() ? I32.changeVectorElementType(MVT::i16) : MVT::i16;
+    EVT I16 = I32.isVector()
```
Can we use `changeElementType` here?

topperc commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -4215,7 +4216,8 @@ SDValue AMDGPUTargetLowering::performSraCombine(SDNode *N,
   EVT ElementType = VT.getScalarType();
   EVT TargetScalarType = ElementType.getHalfSizedIntegerVT(*DAG.getContext());
-  EVT TargetType = VT.isVector() ? VT.changeVectorElementType(TargetScalarType)
+  EVT TargetType = VT.isVector() ? VT.changeVectorElementType(*DAG.getContext(),
```
Can we use `changeElementType` here?

topperc commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -4337,7 +4339,8 @@ SDValue AMDGPUTargetLowering::performSrlCombine(SDNode *N,
   EVT ElementType = VT.getScalarType();
   EVT TargetScalarType = ElementType.getHalfSizedIntegerVT(*DAG.getContext());
-  EVT TargetType = VT.isVector() ? VT.changeVectorElementType(TargetScalarType)
+  EVT TargetType = VT.isVector() ? VT.changeVectorElementType(*DAG.getContext(),
```
Can we use `changeElementType` here?

topperc commented on llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:
```diff
@@ -2514,8 +2515,10 @@ SDValue NVPTXTargetLowering::LowerFP_EXTEND(SDValue Op,
         (STI.getSmVersion() < 90 || STI.getPTXVersion() < 78)) {
-      EVT F32 = NarrowVT.isVector() ? NarrowVT.changeVectorElementType(MVT::f32)
-                                    : MVT::f32;
+      EVT F32 =
```
Can we use `changeElementType` here?

topperc commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -34424,7 +34424,9 @@ void X86TargetLowering::ReplaceNodeResults(SDNode *N,
     SDValue Res;
     if (isSoftF16(SrcVT, Subtarget)) {
-      EVT NVT = VT.isVector() ? VT.changeVectorElementType(MVT::f32) : MVT::f32;
+      EVT NVT = VT.isVector()
```
Can we use `changeElementType` here?

Islam-Imad committed:
[SelectionDAG] Simplify EVT construction using changeElementType

This patch simplifies EVT type construction across the codebase by
utilizing the changeElementType() and changeVectorElementType() helper
functions instead of manual conditional logic.

Islam-Imad commented:
I have refactored EVT construction to use changeElementType() and changeVectorElementType() functions instead of manual conditional logic.
is there something missing i should do ?

topperc reviewed (approved):
LGTM

Islam-Imad committed:
Merge branch 'main' into fix/evt-change-element-type-extended-fallback

Islam-Imad commented:
> I have refactored EVT construction to use changeElementType() and changeVectorElementType() functions instead of manual conditional logic. is there something missing i should do ?

@RKSimon :)

Islam-Imad committed:
Merge branch 'main' into fix/evt-change-element-type-extended-fallback

RKSimon reviewed (commented):
cheers - just noticed one minor

Islam-Imad committed:
Refactor redundant function calls

RKSimon reviewed (approved):
LGTM

RKSimon committed:
Merge branch 'main' into fix/evt-change-element-type-extended-fallback
